### DISCUSSION TITLE
Is `required`, `[Required]`, and `[JsonRequired]` equivalent in `Text.Json`?

### DISCUSSION THREAD

**Discussion by ptr727** [dotnet/runtime#100555] [Q&A]
I'm converting code from newtonsoft.json to text.json, and I'd like to get clarity on "required" attributes/fields?

Is C#11 `required` modifier the same as data annotations `[Required]`?  
Does .NET8 Text.Json need `[JsonRequired]` if `[Required]` is already set?

**Comment by DaZombieKiller** [comment]
`JsonRequiredAttribute` should be equivalent to `required` (but the latter should probably be preferred when possible). When you mark something with `required`, at the IL level it gets marked with `System.Runtime.CompilerServices.RequiredMemberAttribute`. You can see where System.Text.Json determines if a member is required [here](https://github.com/dotnet/runtime/blob/59d9749a225d27ef27aaa039ff561f3c3c7a6b83/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/DefaultJsonTypeInfoResolver.Helpers.cs#L404-L409). It first checks for `JsonRequiredAttribute`, and if that is not present it checks for `RequiredMemberAttribute`. `System.ComponentModel.DataAnnotations.RequiredAttribute` is unrelated and does not appear to be acknowledged by System.Text.Json.

**Reply by ptr727** [reply]
Thank you. If `DataAnnotations` are considered "unrelated", is there an equivalent in C#11 or Text.Json for `[Obsolete]`?  
To be specific, I want the attribute to be read, e.g. old schema config, but not written, i.e. current config.

I am currently doing it using:

```csharp
    private static void ExcludeObsoletePropertiesModifier(JsonTypeInfo typeInfo)
    {
        // Only process objects
        if (typeInfo.Kind != JsonTypeInfoKind.Object)
            return;

        // Iterate over all properties
        foreach (var property in typeInfo.Properties)
        {
            // Do not serialize [Obsolete] items
            if (property.AttributeProvider?.IsDefined(typeof(ObsoleteAttribute), true) == true)
                property.ShouldSerialize = (_, _) => false;
        }
    }
```
